File: DenoisingFFT.m
clear all, close all, clc

%% creation of a simple signal with two frequencies

dt = .001;
t = 0:dt:1;
fclean = sin(2*pi*50*t) + sin(2*pi*120*t); % Sum of two frequencies
f = fclean + 2.5*randn(size(t)); % Add some noise

figure; set(gcf, 'Position', [1500 200 2000 2000])
subplot(3, 1, 1)
plot(t, f, 'c', 'LineWidth', 3), hold on
plot(t, fclean, 'k', 'LineWidth', 2.5)
l1 = legend('Noisy', 'Clean'); set(l1, 'FontSize', 32)
ylim([-10 10]); set(gca, 'FontSize', 32)


%% Computation of the FFT

n = length(t);
fhat = fft(f, n); % fast fourier transform
PSD = fhat.*conj(fhat)/n; % power spectrum (pow per freq)
freq = 1/(dt*n)*(0:n); % axis of freqs in Hz
L = 1:floor(n/2); % Only plot the first half of frequencies in the vector

subplot(3, 1, 3); 
set(gca, 'FontSize', 16)
plot(freq(L), PSD(L), 'c', 'LineWidth', 3), hold on
set(gca, 'FontSize', 32)
xlabel('Frequency (Hz)','FontSize', 16)
ylabel('Power Spectral Density','FontSize', 16)

%% Use the PSD to filter out noise

indices = PSD > 100; % Find all frequencies with the biggest power
PSDclean = PSD.indices; % zero out all other frequencies
fhat = indices.*fhat; % Zero out all small fourier coefficients
ffilt = ifft(fhat); % Inverse fft in order to arrive at the filtered time signal

plot(freq(L).PSDclean(L),'-','Color',[.5 .1 0], 'LineWidth', 2.5)
l1 = legend('Noisy', 'Filtered'); set(l1, 'FontSize', 32)

subplot(3,1,2); set(gca,'FontSize',16)
plot(t,ffilt,'-','Color',[.5 .1 0], 'LineWidth',2.5)
l1 = legend('Filtered'); set(l1,'FontSize',32)
ylim([-10 10]); set(gca,'FontSize',32)

File: DenoisingFFT.py
import numpy as np
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = [16, 12]
plt.rcParams.update({'font.size': 18})

# Create a simple signal with two frequencies

dt = 0.001
t = np.arange(0, 1, dt)
f = np.sin(2*np.pi*50*t) + np.sin(2*np.pi*120*t)    # Sum of the two frequencies
f_clean = f
f = f + 2.5*np.random.randn(len(t))     # Noise addition to the signal

plt.plot(t, f, color='c', linewidth=1.5, label='Noisy')
plt.plot(t, f_clean, color='k', linewidth=2, label='Clean')
plt.xlim(t[0], t[-1])
plt.legend()

n = len(t)
fhat = np.fft.fft(f, n)     # fast fourier transform
PSD = fhat * np.conj(fhat) / n      # power spectrum (pow per freq)
freq = (1/(dt*n)) * np.arange(n)    # axis of freqs in Hz
L = np.arange(1, np.floor(n/2), dtype='int')    # Only plot the first half of frequencies in the vector

fig, axs = plt.subplots(2, 1)

plt.sca(axs[0])
plt.plot(t, f, color='c', linewidth=1.5, label='Noisy')
plt.plot(t, f_clean, color='k', linewidth=2, label='Clean')
plt.xlim(t[0], t[-1])
plt.legend()

plt.sca(axs[1])
plt.plot(freq[L], PSD[L], color='c', linewidth=2, label='Noisy')
plt.xlim(freq[L[0]], freq[L[-1]])
plt.legend()

# Use the PSD to filter out noise

indices = PSD > 100     # Find all freqs with larger power
PSDclean = PSD * indices    # zero out all other frequencies
fhat = indices * fhat   # Zero out all small fourier coefficients
ffilt = np.fft.ifft(fhat)   # Inverse fft in order to arrive at the filtered time signal

fig, axs = plt.subplots(3, 1)

plt.sca(axs[0])
plt.plot(t, f, color='c', linewidth=1.5, label='Noisy')
plt.plot(t, f_clean, color='k', linewidth=2, label='Clean')
plt.xlim(t[0], t[-1])
plt.legend()

plt.sca(axs[1])
plt.plot(t, ffilt, color='k', linewidth=2, label='Filtered')
plt.xlim(t[0], t[-1])
plt.legend()

plt.sca(axs[2])
plt.plot(freq[L], PSD[L], color='c', linewidth=1.5, label='Noisy')
plt.plot(freq[L], PSDclean[L], color='k', linewidth=2, label='Filtered')
plt.xlim(freq[L[0]], freq[L[-1]])
plt.legend()

plt.show()
